File: python/unified_channel/adapters/discord.py
"""Discord adapter — bridges discord.py to UnifiedMessage."""

from __future__ import annotations

import asyncio
import logging
from datetime import datetime
from typing import AsyncIterator

import discord

from ..adapter import ChannelAdapter
from ..types import (
    ChannelStatus,
    ContentType,
    Identity,
    MessageContent,
    OutboundMessage,
    UnifiedMessage,
)

logger = logging.getLogger(__name__)


class DiscordAdapter(ChannelAdapter):
    """
    Discord channel adapter using discord.py.

    Only responds in specified channel IDs (or DMs if allow_dm=True).
    """

    channel_id = "discord"

    def __init__(
        self,
        token: str,
        *,
        allowed_channel_ids: set[int] | None = None,
        allow_dm: bool = True,
        command_prefix: str = "/",
    ) -> None:
        self._token = token
        self._allowed_channels = allowed_channel_ids
        self._allow_dm = allow_dm
        self._prefix = command_prefix
        self._queue: asyncio.Queue[UnifiedMessage] = asyncio.Queue()
        self._connected = False
        self._last_activity: datetime | None = None
        self._bot_user: str | None = None

        intents = discord.Intents.default()
        intents.message_content = True
        self._client = discord.Client(intents=intents)
        self._setup_handlers()

    def _setup_handlers(self) -> None:
        @self._client.event
        async def on_ready():
            self._connected = True
            self._bot_user = str(self._client.user)
            logger.info("discord connected: %s", self._bot_user)

        @self._client.event
        async def on_message(message: discord.Message):
            # Ignore own messages
            if message.author == self._client.user:
                return

            # Filter channels
            if isinstance(message.channel, discord.DMChannel):
                if not self._allow_dm:
                    return
            elif self._allowed_channels and message.channel.id not in self._allowed_channels:
                return

            content = message.content or ""
            self._last_activity = datetime.now()

            # Check if it's a command
            if content.startswith(self._prefix):
                parts = content[len(self._prefix):].split()
                cmd = parts[0] if parts else ""
                args = parts[1:]
                mc = MessageContent(
                    type=ContentType.COMMAND,
                    text=content,
                    command=cmd,
                    args=args,
                )
            elif message.attachments:
                mc = MessageContent(
                    type=ContentType.MEDIA,
                    text=content,
                    media_url=message.attachments[0].url,
                    media_type=message.attachments[0].content_type or "unknown",
                )
            else:
                mc = MessageContent(type=ContentType.TEXT, text=content)

            msg = UnifiedMessage(
                id=str(message.id),
                channel="discord",
                sender=Identity(
                    id=str(message.author.id),
                    username=message.author.name,
                    display_name=message.author.display_name,
                ),
                content=mc,
                timestamp=message.created_at,
                chat_id=str(message.channel.id),
                thread_id=(
                    str(message.thread.id)
                    if hasattr(message, "thread") and message.thread
                    else None
                ),
                reply_to_id=(
                    str(message.reference.message_id)
                    if message.reference
                    else None
                ),
                raw=message,
            )
            await self._queue.put(msg)

    async def connect(self) -> None:
        # Start client in background — discord.py's start() does not block
        asyncio.create_task(self._client.start(self._token))
        # Wait for ready
        for _ in range(30):
            if self._connected:
                return
            await asyncio.sleep(1)
        raise RuntimeError("discord connection timed out")

    async def disconnect(self) -> None:
        await self._client.close()
        self._connected = False
        logger.info("discord disconnected")

    async def receive(self) -> AsyncIterator[UnifiedMessage]:
        while self._connected:
            try:
                msg = await asyncio.wait_for(self._queue.get(), timeout=1.0)
                yield msg
            except asyncio.TimeoutError:
                continue

    async def send(self, msg: OutboundMessage) -> str | None:
        channel = self._client.get_channel(int(msg.chat_id))
        if not channel:
            # Try as DM user ID
            try:
                user = await self._client.fetch_user(int(msg.chat_id))
                channel = await user.create_dm()
            except Exception:
                logger.error("discord: cannot find channel/user %s", msg.chat_id)
                return None

        kwargs: dict = {"content": msg.text}
        if msg.reply_to_id:
            try:
                ref = await channel.fetch_message(int(msg.reply_to_id))  # type: ignore
                kwargs["reference"] = ref
            except Exception:
                pass

        sent = await channel.send(**kwargs)  # type: ignore
        self._last_activity = datetime.now()
        return str(sent.id)

    async def get_status(self) -> ChannelStatus:
        return ChannelStatus(
            connected=self._connected,
            channel="discord",
            account_id=self._bot_user,
            last_activity=self._last_activity,
        )

File: rust/src/adapters/discord.rs
//! Discord adapter using the serenity crate.

use async_trait::async_trait;
use chrono::Utc;
use std::collections::HashSet;
use std::sync::Arc;
use tokio::sync::Mutex;

use serenity::all::{
    ChannelId, Context, CreateMessage, EventHandler, GatewayIntents, Message, Ready,
};
use serenity::Client;

use crate::adapter::{ChannelAdapter, MessageHandler};
use crate::error::{Error, Result};
use crate::types::*;

/// Configuration for the Discord adapter.
pub struct DiscordOptions {
    /// Restrict to specific channel IDs. None means all channels.
    pub allowed_channel_ids: Option<HashSet<String>>,
    /// Allow direct messages.
    pub allow_dm: bool,
    /// Command prefix (default "/").
    pub command_prefix: String,
}

impl Default for DiscordOptions {
    fn default() -> Self {
        Self {
            allowed_channel_ids: None,
            allow_dm: true,
            command_prefix: "/".to_string(),
        }
    }
}

/// Serenity event handler that forwards to the unified handler.
struct DiscordEventHandler {
    handler: Arc<Mutex<Option<MessageHandler>>>,
    options: Arc<DiscordOptions>,
}

#[async_trait]
impl EventHandler for DiscordEventHandler {
    async fn message(&self, _ctx: Context, msg: Message) {
        // Skip bot messages
        if msg.author.bot {
            return;
        }

        // DM check
        if msg.is_private() && !self.options.allow_dm {
            return;
        }

        // Channel allowlist
        if !msg.is_private() {
            if let Some(ref allowed) = self.options.allowed_channel_ids {
                if !allowed.contains(&msg.channel_id.to_string()) {
                    return;
                }
            }
        }

        let text = msg.content.clone();
        let prefix = &self.options.command_prefix;
        let is_cmd = text.starts_with(prefix);

        let content = if is_cmd {
            let stripped = &text[prefix.len()..];
            let parts: Vec<&str> = stripped.split_whitespace().collect();
            let cmd = parts.first().unwrap_or(&"").to_string();
            let args: Vec<String> = parts[1..].iter().map(|s| s.to_string()).collect();
            MessageContent::command(text, cmd, args)
        } else if !msg.attachments.is_empty() {
            let att = &msg.attachments[0];
            MessageContent::media(
                text,
                att.url.clone(),
                att.content_type.clone(),
            )
        } else {
            MessageContent::text(text)
        };

        let sender = Identity::new(msg.author.id.to_string())
            .with_username(msg.author.name.clone())
            .with_display_name(msg.author.global_name.clone().unwrap_or(msg.author.name.clone()));

        let unified = UnifiedMessage {
            id: msg.id.to_string(),
            channel: "discord".to_string(),
            sender,
            content,
            timestamp: *msg.timestamp,
            thread_id: None,
            reply_to_id: msg.referenced_message.as_ref().map(|r| r.id.to_string()),
            chat_id: Some(msg.channel_id.to_string()),
            raw: None,
            metadata: Default::default(),
        };

        let handler = self.handler.lock().await;
        if let Some(ref h) = *handler {
            h(unified);
        }
    }

    async fn ready(&self, _ctx: Context, ready: Ready) {
        tracing::info!(user = %ready.user.name, "discord connected");
    }
}

/// Discord adapter backed by serenity.
pub struct DiscordAdapter {
    token: String,
    options: Arc<DiscordOptions>,
    connected: bool,
    last_activity: Option<chrono::DateTime<Utc>>,
    bot_user: Option<String>,
    handler: Arc<Mutex<Option<MessageHandler>>>,
    client: Option<Arc<Mutex<Client>>>,
}

impl DiscordAdapter {
    /// Create a new Discord adapter.
    pub fn new(token: impl Into<String>) -> Self {
        Self {
            token: token.into(),
            options: Arc::new(DiscordOptions::default()),
            connected: false,
            last_activity: None,
            bot_user: None,
            handler: Arc::new(Mutex::new(None)),
            client: None,
        }
    }

    /// Builder: set options.
    pub fn with_options(mut self, options: DiscordOptions) -> Self {
        self.options = Arc::new(options);
        self
    }
}

#[async_trait]
impl ChannelAdapter for DiscordAdapter {
    fn channel_id(&self) -> &str {
        "discord"
    }

    async fn connect(&mut self) -> Result<()> {
        let intents = GatewayIntents::GUILD_MESSAGES
            | GatewayIntents::MESSAGE_CONTENT
            | GatewayIntents::DIRECT_MESSAGES;

        let event_handler = DiscordEventHandler {
            handler: Arc::clone(&self.handler),
            options: Arc::clone(&self.options),
        };

        let client = Client::builder(&self.token, intents)
            .event_handler(event_handler)
            .await
            .map_err(|e| Error::Connection(e.to_string()))?;

        let client = Arc::new(Mutex::new(client));
        self.client = Some(Arc::clone(&client));

        // Start in background
        tokio::spawn(async move {
            let mut c = client.lock().await;
            if let Err(e) = c.start().await {
                tracing::error!(error = %e, "discord client error");
            }
        });

        self.connected = true;
        Ok(())
    }

    async fn disconnect(&mut self) -> Result<()> {
        if let Some(ref client) = self.client {
            let mut c = client.lock().await;
            c.shard_manager.shutdown_all().await;
        }
        self.connected = false;
        Ok(())
    }

    fn on_message(&mut self, handler: MessageHandler) {
        // We can't easily do async set here, so we use a blocking approach
        // since on_message is called before connect.
        let h = Arc::clone(&self.handler);
        tokio::task::block_in_place(|| {
            let rt = tokio::runtime::Handle::current();
            rt.block_on(async {
                let mut guard = h.lock().await;
                *guard = Some(handler);
            });
        });
    }

    async fn send(&self, msg: OutboundMessage) -> Result<Option<String>> {
        let client = self.client.as_ref().ok_or_else(|| {
            Error::Send("not connected".to_string())
        })?;

        let channel_id: u64 = msg
            .chat_id
            .parse()
            .map_err(|_| Error::Send(format!("invalid channel_id: {}", msg.chat_id)))?;

        let c = client.lock().await;
        let http = c.http.clone();
        drop(c);

        let channel = ChannelId::new(channel_id);
        let builder = CreateMessage::new().content(&msg.text);
        let sent = channel
            .send_message(&http, builder)
            .await
            .map_err(|e| Error::Send(e.to_string()))?;

        Ok(Some(sent.id.to_string()))
    }

    async fn get_status(&self) -> ChannelStatus {
        ChannelStatus {
            connected: self.connected,
            channel: "discord".to_string(),
            account_id: self.bot_user.clone(),
            error: None,
            last_activity: self.last_activity,
        }
    }
}
